### PR TITLE
T/ckeditor5/1096: Introduce schema.isInline definition.

### PR DESCRIPTION
**Pull request by jodator**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Introduce inline object type to schema. Closes ckeditor/ckeditor5#1049. Closes ckeditor/ckeditor5#1426.

---

### Additional information

* Part of changes required for inline-widgets. [Master PR](https://github.com/ckeditor/ckeditor5-widget/pull/70).

### CODE REVIEW

File: src/conversion/mapper-helpers.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+		}
+	};
+}
+
+/**
+ * Maps model position to view position if the position set after/before inline element (ie. inline widget).
+ *
+ *		editor.editing.mapper.on( 'modelToView', mapModelPositionOnInlineElement( model, view ) );
+ *
+ * This methods ensures that selection set on inline element is set outside surrounding text nodes as by default the position would be set
```
Comment (ma2ciek) [inline src/conversion/mapper-helpers.js]:
```suggestion
 * This method ensures that selection set on an inline element is set outside surrounding text nodes as by default the position would be set
```

File: src/conversion/mapper-helpers.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+ * at the end or begging of a text node that is previous/next to the inline element:
```
Comment (ma2ciek) [inline src/conversion/mapper-helpers.js]:
```suggestion
 * at the end or beginning of a text node that is previous/next to the inline element:
```

**Comment by Reinmar** [comment]
> Feature: Introduce inline object type to schema. Closes [ckeditor/ckeditor5#1049](https://github.com/ckeditor/ckeditor5/issues/1049). Closes [ckeditor/ckeditor5#1426](https://github.com/ckeditor/ckeditor5/issues/1426).

U sure that this description covers the changes? ;)

**Comment by jodator** [comment]
Well it does introduces the `isInline` thingy to the schema with additional changes needed to make that work. It also changes `$text` schema definition to include it (dunno if BREAKING CHANGE though).

> Feature: Introduce inline object type to schema.

But I don't know how to name it nicely without writing an elaboration about that.

Feature: Introduce inline object type to the editor model schema, handling positions around inline object elements and add isInline=ture to $text schema definition.

Would probably cover that better but is too long...

**Comment by Reinmar** [comment]
What about?

image

**Comment by jodator** [comment]
This is outcome of introducing the inline elements. But i get the point that it also needs a mention. I'll try to came with something better.

**Comment by Reinmar** [comment]
I think we can do better with mapping positions than registering custom converters. @jodator explained to me that those converters were added by him because view position that was located inside an inline element was not correctly converted by the mapper which was crashing the editor. The converters proposed by @jodator handle those cases as well as try to do what selection post-fixer would normally do.

I think we should fix mapper so it's able to convert positions which end up inside inline model elements. Those positions are still fine and, ideally, should work and not crash the editor. In the future, this will handle well inline widgets with inline nested editables.

Then, we need to figure out whether the selection post-fixer is correcting those selections that start outside an inline limit element and end inside it – just like in the block widget cases.

**Comment by scofalik** [comment]
I've pushed two changes: removed mapping callbacks and fixed `view.Selection#getSelectedElement()`. I hope this covers what we discussed F2F. I will also publish the removed mapping callbacks as utils in `ckeditor5-widget` repo.

**Pull request merged**